File: 0001-two-sum/0001-two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
         // Brute Force

        // for(int i = 0; i < nums.length; i++) {
        //     for(int j = i+1; j < nums.length; j++) {
        //         if(nums[i] + nums[j] == target) return new int[] {i, j};
        //     }
        // }

        // return new int[] {};

        // HashMap

        Map<Integer, Integer> map = new HashMap<>();

        for(int i = 0; i < nums.length; i++) {
            int remain = (target - nums[i]);

            if(map.containsKey(remain)) return new int[] {map.get(remain), i};

            map.put(nums[i], i);
        }

        return new int[] {};
    }
}

File: 0001-two-sum/0001-two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # Brute Force Approch

        # n = len(nums)

        # for i in range(n) :
        #     for j in range(i+1, n) :
        #         if (nums[i] + nums[j]) == target :
        #             return list([i, j])
        
        # return list()

        # Optimal Solution 

        n = len(nums)
        freq_map = {}

        for i in range(n) :
            x = target - nums[i]

            if x in freq_map :
                return list([freq_map[x], i])
            else :
                freq_map[nums[i]] = i


        return list()